### PR TITLE
Add quiz for chapter 8 (Generic Types and Traits)

### PR DESCRIPTION
**Pull request by CollinsC1O**
This change is [Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/785)

### CODE REVIEW

File: quizzes/ch08-generic-types-and-traits.toml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+[[questions]]
+type = "short_answer"
+prompt.prompt = "What keyword is used in Cairo to define a generic type in a function signature?"
```
Comment (enitrat) [inline quizzes/ch08-generic-types-and-traits.toml:3]: This is not correct, you can use any placeholder type, `T` is only a convention - but when you have multiple generic types you can also use other letters, such as in the `Result<T, E> ` enum. 

Moreover the question is of low interest, see #729 to see what constitutes a good and bad quiz

File: quizzes/ch08-generic-types-and-traits.toml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+context = """
+    Generics are abstract stand-ins for concrete types or other properties, 
+    allowing for the creation of reusable code that can operate on many different types, 
+    thus avoiding code duplication and enhancing code maintainability.
+"""
+
+[[questions]]
+type = "MultipleChoice"
+prompt.prompt = """
+How does the 'main' function below determine the largest number in the provided array?
```
Comment (enitrat) [inline quizzes/ch08-generic-types-and-traits.toml:31]: This question has nothing to do with generics 🤔

**Review by enitrat — COMMENTED** [review]
Thank you for the contribution, however I don't think the topic has been very well understood, the questions are not challenging and do not engage a lot of reflection. There should be more questions with code examples and common "traps" that developers might fall into initially.

File: quizzes/ch08-generic-types-and-traits.toml (hunk 2 above)
Comment (CollinsC1O) [reply quizzes/ch08-generic-types-and-traits.toml:31]: oh that's true. I will have to implement a correct question

File: quizzes/ch08-generic-types-and-traits.toml (hunk 1 above)
Comment (CollinsC1O) [reply quizzes/ch08-generic-types-and-traits.toml:3]: Thank you for your assistance. I will now proceed to implement the necessary changes

Comment (CollinsC1O) [reply quizzes/ch08-generic-types-and-traits.toml]: Hi @enitrat I have implemented the changes please review

Comment (enitrat) [reply quizzes/ch08-generic-types-and-traits.toml]: @CollinsC1O I still don't see any concrete questions with examples

Comment (CollinsC1O) [reply quizzes/ch08-generic-types-and-traits.toml]: ok I will have to do better

**Review by enitrat — COMMENTED** [review]
Hi, thank you for your time, but ss I'm still not seeing the changes requested and concrete examples, I will close this in favor of another implementation.

**Comment by CollinsC1O** [comment]
Hi @enitrat please don't close it yet, I'm currently working on the three chapters and trying my best to effect the requested changes. I want to give a very good implementation

**Comment by CollinsC1O** [comment]
I really want to implement this well

**Comment by enitrat** [comment]
Ok, then please focus on giving both concrete examples with questions leading to reflexion, and theory questions as wells.

For reflexion:
- What happens when you have defined methods on a type T, and you implement a trait that contains a function with the same signature?
- Given `fn foo<T>(x:T) -> T;` what is the only content possible for the function `foo`?
- Can you define a function that takes an Array<T> and prints all its elements?

File: quizzes/ch08-01-generic-data-types.toml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+type = "ShortAnswer"
+prompt.prompt = """
+How would you define a generic function that takes any type of integer 
+and returns its square? Provide a Cairo code snippet as an example.
+"""
+answer.answer = """
+You can define a generic function in Cairo using the `fn` keyword followed by the function 
+name and generic type parameters in angle brackets. For example:
+```
+fn square<T>(x: T) -> T {
```
Comment (enitrat) [inline quizzes/ch08-01-generic-data-types.toml:11]: Have you tried this?

File: quizzes/ch08-01-generic-data-types.toml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    return x * x;
+}
+```
+"""
+
+[[questions]]
+type = "Tracing"
+prompt.prompt = """
+fn largest<T> (arr1: Array<T>, arr2: Array<T>) -> Array<T> {
```
Comment (enitrat) [inline quizzes/ch08-01-generic-data-types.toml:20]: Have you tried this?

File: quizzes/ch08-01-generic-data-types.toml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+fn main() {
+    let mut arr1 = array![2, 1, 5, 4, 9];
+    let mut arr2 = array![2, 1, 5, 4, 9, 3, 1];
+
+    let largestArr = largest(arr1, arr2);
+}
+"""
+answer.doesCompile = false
+context = "The code won't compile due to the compiler's inability to ensure that an Array<T>
```
Comment (enitrat) [inline quizzes/ch08-01-generic-data-types.toml:36]: This is not the (only) reason why it won't compile

File: quizzes/ch08-01-generic-data-types.toml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    let mut age = studentI.age;
+    let mut level = studentI.level;
+    
+    println!("{}", level);
+}
+"""
+answer.doesCompile = true
+answer.stdout = "9"
+
```
Comment (enitrat) [inline quizzes/ch08-01-generic-data-types.toml:60]: Missing context: why does it compile? (`name` is inferred as a `felt252`, so is the rest)

File: quizzes/ch08-01-generic-data-types.toml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+[[questions]]
+type = "Tracing"
+prompt.prompt = """
+#[derive(Drop)]
```
Comment (enitrat) [inline quizzes/ch08-01-generic-data-types.toml:64]: I really don't see the purpose of this question

File: quizzes/ch08-01-generic-data-types.toml (hunk 4 above)
Comment (CollinsC1O) [reply quizzes/ch08-01-generic-data-types.toml:20]: Yes I tried it out. It compiles only when the 'drop' trait is implemented, but it won't compile due to the necessity of dropping the variables 'arr1' and 'arr2'.

File: quizzes/ch08-01-generic-data-types.toml (hunk 6 above)
Comment (CollinsC1O) [reply quizzes/ch08-01-generic-data-types.toml:60]: Oh, I'll need to include that. This compiles for several reasons, including accurate trait implementations, proper function calls and returns, correct memory management, proper variable scope handling, and proper type declarations and usage and some other reasons. I wanted to share with you a few things I know before implementing the changes.

File: quizzes/ch08-01-generic-data-types.toml (hunk 5 above)
Comment (CollinsC1O) [reply quizzes/ch08-01-generic-data-types.toml:36]: One of the main reasons for the compilation failure is the lack of 'drop' trait implementation in the function signature. Consequently, when executing 'scarb build' or 'scarb cairo-run', an error message "variable not dropped" will be displayed. However, once the 'drop' trait is declared in the placeholder, just after T just as: '<T, impl TDrot: Drop<T>', the program will compile. I've tried it.

File: quizzes/ch08-01-generic-data-types.toml (hunk 7 above)
Comment (CollinsC1O) [reply quizzes/ch08-01-generic-data-types.toml:64]: Ok I will have to remove it or change it completely.
I'm very grateful for your patience, I really have to do this right. I'm truly learning allot

File: quizzes/ch08-01-generic-data-types.toml (hunk 4 above)
Comment (enitrat) [reply quizzes/ch08-01-generic-data-types.toml:20]: That's not the only reason why it wont compile!

File: quizzes/ch08-01-generic-data-types.toml (hunk 7 above)
Comment (enitrat) [reply quizzes/ch08-01-generic-data-types.toml:64]: sure, take your time, we're getting there :)

**Review by enitrat — APPROVED** [review]

**Review by enitrat — APPROVED** [review]

**Comment by CollinsC1O** [comment]
Hello @enitrat, I'm very grateful for the opportunity to have worked under you. This means a lot to me. Thank you for the merge. I would love to follow up with you on X and Telegram if you don't mind.

**Pull request merged**